servers/src/main/java/tachyon/worker/block/evictor/EvictorBase.java

```diff
@@ -0,0 +1,204 @@
+        try {
+          BlockMeta block = mManagerView.getBlockMeta(blockId);
+          if (null == block) {
+            continue;
+          }
+          StorageDirView nextDirView =
+              mAllocator.allocateBlockWithView(Users.MIGRATE_DATA_USER_ID, block.getBlockSize(),
+                  BlockStoreLocation.anyDirInTier(nextTierView.getTierViewAlias()), mManagerView);
+          if (nextDirView == null) {
+            nextDirView =
+                cascadingEvict(block.getBlockSize(),
+                    BlockStoreLocation.anyDirInTier(nextTierView.getTierViewAlias()), plan);
+          }
+          if (nextDirView == null) {
+            // If we failed to find a dir in the next tier to move this block, evict it and
+            // continue. Normally this should not happen.
+            plan.toEvict().add(blockId);
+            candidateDirView.markBlockMoveOut(blockId, block.getBlockSize());
+            continue;
+          }
+          plan.toMove().add(
+              new Pair<Long, BlockStoreLocation>(blockId, nextDirView.toBlockStoreLocation()));
+          candidateDirView.markBlockMoveOut(blockId, block.getBlockSize());
+          nextDirView.markBlockMoveIn(blockId, block.getBlockSize());
+        } catch (NotFoundException nfe) {
+          continue;
+        }
+      }
+    }
+
+    return candidateDirView;
+  }
+
+  @Override
+  public EvictionPlan freeSpaceWithView(long bytesToBeAvailable, BlockStoreLocation location,
+      BlockMetadataManagerView view) {
+    mManagerView = view;
+
+    List<Pair<Long, BlockStoreLocation>> toMove = new ArrayList<Pair<Long, BlockStoreLocation>>();
+    List<Long> toEvict = new ArrayList<Long>();
+    EvictionPlan plan = new EvictionPlan(toMove, toEvict);
+    StorageDirView candidateDir = cascadingEvict(bytesToBeAvailable, location, plan);
+
+    mManagerView.clearBlockMarks();
+    if (candidateDir == null) {
+      return null;
+    }
+
+    return plan;
+  }
+
+  /**
+   * @return an iterator over the blocks in the evictor cache. The evictor is responsible for
+   * specifying the iteration order using its own strategy. For example, LRUEvictor returns an
+   * iterator that iterates the blocks in LRU order. The key of the map entry is the block Id.
+   */
+  protected abstract Iterator<Map.Entry<Long, Object>> getIterator();
+
+  /**
+   * Perform additional cleanup when a block is removed from the iterator returned by
+   * {@link #getIterator}.
+   */
+  protected void onRemoveBlockFromIterator(long blockId) {}
+
+  /**
+   * Update the block store location if the evictor wants to free space in a specific location.
+   * For example, PartialLRUEvictor always evicts blocks from a dir with max free space.
+   *
+   * @param bytesToBeAvailable bytes to be available after eviction
+   * @param location the original block store location
+   * @return the updated block store location
+   */
+  protected BlockStoreLocation updateBlockStoreLocation(long bytesToBeAvailable,
```

Comment:
Does this mean the eviction policy has higher priority over user input and after eviction runs its possible the allocation attempt will fail?